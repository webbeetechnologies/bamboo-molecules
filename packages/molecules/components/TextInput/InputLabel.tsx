import { memo, useCallback, useMemo, useState } from 'react';
import { Animated, LayoutChangeEvent } from 'react-native';
import { StyleSheet } from 'react-native-unistyles';

import type { InputLabelProps } from './types';

const InputLabel = (props: InputLabelProps) => {
    const {
        parentState,
        style,
        label,
        labelBackground,
        floatingLabelVerticalOffset,
        error,
        onLayoutAnimatedText,
        baseLabelTranslateX,
        wiggleOffsetX,
        labelScale,
        paddingOffset,
        labelTranslationXOffset,
        maxFontSizeMultiplier,
        required,
        testID,
    } = props;

    const [containerLayout, setContainerLayout] = useState<{
        measured: boolean;
        width: number;
        height: number;
    }>({
        measured: false,
        width: 0,
        height: 0,
    });

    const handleLayoutContainer = useCallback((e: LayoutChangeEvent) => {
        setContainerLayout({
            width: e.nativeEvent.layout.width,
            height: e.nativeEvent.layout.height,
            measured: true,
        });
    }, []);

    const { containerStyle, minimizedLabelStyle, normalLabelStyle } = useMemo(() => {
        const isLabelFloating = parentState.value || parentState.focused;

        const labelStyle = {
            transform: [
                {
                    // Wiggle the label when there's an error
                    translateX: parentState.errorAnimation.interpolate({
                        inputRange: [0, 0.5, 1],
                        outputRange: [0, parentState.value && error ? wiggleOffsetX : 0, 0],
                    }),
                },
                {
                    // Move label to top
                    translateY: parentState.labelAnimation.interpolate({
                        inputRange: [0, 1],
                        outputRange: [
                            (floatingLabelVerticalOffset || 0) - containerLayout.height / 2,
                            0,
                        ],
                    }),
                },
                {
                    // Make label smaller
                    scale: parentState.labelAnimation.interpolate({
                        inputRange: [0, 1],
                        outputRange: [labelScale, 1],
                    }),
                },
            ],
        };
        const labelTranslationX = {
            transform: [
                {
                    // Offset label scale since RN doesn't support transform origin
                    translateX: parentState.labelAnimation.interpolate({
                        inputRange: [0, 1],
                        outputRange: [baseLabelTranslateX, labelTranslationXOffset || 0],
                    }),
                },
            ],
        };

        return {
            containerStyle: [
                StyleSheet.absoluteFill,
                styles.labelContainer,
                {
                    opacity:
                        // Hide the label in minimized state until we measure it's width
                        !isLabelFloating || parentState.labelLayout.measured ? 1 : 0,
                },
                labelTranslationX,
            ],
            minimizedLabelStyle: [
                // {
                //     color: theme.colors.onSurface,
                // },
                labelStyle,
                paddingOffset || {},
                {
                    opacity: parentState.labelAnimation.interpolate({
                        inputRange: [0, 1],
                        outputRange: [isLabelFloating ? 1 : 0, 0],
                    }),
                },
                style,
                isLabelFloating ? { backgroundColor: labelBackground } : {},
            ],
            normalLabelStyle: [
                // {
                //     color: theme.colors.onSurface,
                // },
                labelStyle,
                paddingOffset || {},
                {
                    opacity: isLabelFloating ? 0 : 1,
                },
                style,
            ],
        };
    }, [
        parentState.value,
        parentState.focused,
        parentState.errorAnimation,
        parentState.labelAnimation,
        parentState.labelLayout.measured,
        error,
        wiggleOffsetX,
        floatingLabelVerticalOffset,
        containerLayout.height,
        labelScale,
        baseLabelTranslateX,
        labelTranslationXOffset,
        // theme.colors.onSurface,
        paddingOffset,
        style,
        labelBackground,
    ]);

    // Position colored placeholder and gray placeholder on top of each other and crossfade them
    // This gives the effect of animating the color, but allows us to use native driver
    return (
        <>
            {label && (
                <Animated.View
                    pointerEvents="none"
                    style={containerStyle}
                    onLayout={handleLayoutContainer}>
                    <Animated.Text
                        onLayout={onLayoutAnimatedText}
                        style={minimizedLabelStyle}
                        numberOfLines={1}
                        maxFontSizeMultiplier={maxFontSizeMultiplier}
                        testID={`${testID}-label-active`}>
                        {label}
                    </Animated.Text>
                    <Animated.Text
                        // variant={parentState.focused ? 'bodyLarge' : 'bodySmall'}
                        style={normalLabelStyle}
                        numberOfLines={1}
                        maxFontSizeMultiplier={maxFontSizeMultiplier}
                        testID={`${testID}-label-inactive`}>
                        {typeof label === 'string' ? `${label}${required ? '*' : ''}` : label}
                    </Animated.Text>
                </Animated.View>
            )}
        </>
    );
};

const styles = StyleSheet.create({
    labelContainer: {
        zIndex: 3,
        justifyContent: 'center',
    },
});

export default memo(InputLabel);
