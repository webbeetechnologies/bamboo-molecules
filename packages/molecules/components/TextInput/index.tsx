import { getRegisteredComponentWithFallback, registerMoleculesComponents } from '../../core';
import TextInputDefault from './TextInput';

registerMoleculesComponents({
    TextInput: TextInputDefault,
});

export const TextInput = getRegisteredComponentWithFallback('TextInput', TextInputDefault);

export {
    Props as TextInputProps,
    ElementProps as TextInputElementProps,
    TextInputHandles,
} from './TextInput';
export { styles as textInputStyles } from './utils';
export type * from './types';
