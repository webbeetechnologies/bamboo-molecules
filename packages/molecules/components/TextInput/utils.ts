import { StyleSheet } from 'react-native-unistyles';

export type States =
    | 'disabled'
    | 'focused'
    | 'hovered'
    | 'hoveredAndFocused'
    | 'errorFocusedAndHovered'
    | 'error'
    | 'errorFocused'
    | 'errorHovered'
    | 'errorDisabled';

export const getInputMinHeight = (variant: string, size: string) => {
    switch (true) {
        case variant === 'outlined' && size === 'lg':
            return 64;
        case variant === 'outlined' && size === 'md':
            return 56;
        case variant === 'outlined' && size === 'sm':
            return 48;
        case variant === 'flat' && size === 'lg':
            return 64;
        case variant === 'flat' && size === 'md':
            return 56;
        case variant === 'flat' && size === 'sm':
            return 52;
        default:
            return 0;
    }
};

export const inputLabelStyles = StyleSheet.create({
    labelContainer: {
        zIndex: 3,
        justifyContent: 'center',
    },
});
