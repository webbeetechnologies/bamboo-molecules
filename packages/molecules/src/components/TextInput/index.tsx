import { getRegisteredMoleculesComponent, registerMoleculesComponents } from '../../core';
import TextInputDefault from './TextInput';

registerMoleculesComponents({
    TextInput: TextInputDefault,
});

export const TextInput = getRegisteredMoleculesComponent('TextInput');

export { Props as TextInputProps, ElementProps as TextInputElementProps } from './TextInput';
export { styles as textInputStyles } from './utils';
