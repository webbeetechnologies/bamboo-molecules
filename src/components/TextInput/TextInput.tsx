import {
    useRef,
    forwardRef,
    useState,
    useEffect,
    useImperativeHandle,
    ReactNode,
    useCallback,
    useMemo,
    memo,
    PropsWithoutRef,
} from 'react';
import {
    Animated,
    TextInput as NativeTextInput,
    LayoutChangeEvent,
    StyleProp,
    TextStyle,
    ViewStyle,
} from 'react-native';
import type { TextInputProps, ViewProps } from '@bambooapp/bamboo-atoms';

import { withActionState, CallbackActionState } from '../../hocs';
import { useComponentStyles, useControlledValue, useMolecules } from '../../hooks';
import type { WithElements } from '../../types';
import TextInputBase from './TextInputBase';
import type { RenderProps, TextInputLabelProp, TextInputSize } from './types';

const BLUR_ANIMATION_DURATION = 180;
const FOCUS_ANIMATION_DURATION = 150;

export type ElementProps = {
    color: string;
    forceFocus: () => void;
    focused: boolean;
};

type Element = ReactNode | ((props: ElementProps) => ReactNode);

export type Props = TextInputProps &
    CallbackActionState &
    WithElements<Element> & {
        /**
         * Variant of the TextInput.
         * - `flat` - flat input with an underline.
         * - `outlined` - input with an outline.
         *
         * In `outlined` variant, the background color of the label is derived from `colors?.background` in theme or the `backgroundColor` style.
         * This component render TextInputOutlined or TextInputFlat based on that props
         */
        variant?: 'flat' | 'outlined' | 'plain';
        /**
         * If true, user won't be able to interact with the component.
         */
        disabled?: boolean;
        /**
         * The text or component to use for the floating label.
         */
        label?: TextInputLabelProp;
        /**
         * Placeholder for the input.
         */
        placeholder?: string;
        /**
         * Whether to style the TextInput with error style.
         */
        error?: boolean;
        /**
         * Callback that is called when the text input's text changes. Changed text is passed as an argument to the callback handler.
         */
        onChangeText?: Function;
        /**
         * Selection color of the input
         */
        selectionColor?: string;
        /**
         * Inactive underline color of the input.
         */
        underlineColor?: string;
        /**
         * Active underline color of the input.
         */
        activeUnderlineColor?: string;
        /**
         * Inactive outline color of the input.
         */
        outlineColor?: string;
        /**
         * Active outline color of the input.
         */
        activeOutlineColor?: string;
        /**
         * Sets min height with densed layout. For `TextInput` in `flat` mode
         * height is `64dp` or in dense layout - `52dp` with label or `40dp` without label.
         * For `TextInput` in `outlined` mode
         * height is `56dp` or in dense layout - `40dp` regardless of label.
         * When you apply `height` prop in style the `dense` prop affects only `paddingVertical` inside `TextInput`
         */
        size?: TextInputSize;
        /**
         * Whether the input can have multiple lines.
         */
        multiline?: boolean;
        /**
         * The number of lines to show in the input (Android only).
         */
        numberOfLines?: number;
        /**
         * The Supporting Text below the TextInput
         */
        supportingText?: string;
        /**
         * To display the required indicator in Supporting Text and in the Label
         */
        required?: boolean;
        /**
         * Callback that is called when the text input is focused.
         */
        onFocus?: (args: any) => void;
        /**
         * Callback that is called when the text input is blurred.
         */
        onBlur?: (args: any) => void;
        /**
         *
         * Callback to render a custom input component such as `react-native-text-input-mask`
         * instead of the default `TextInput` component from `react-native`.
         *
         * Example:
         * ```js
         * <TextInput
         *   label="Phone number"
         *   render={props =>
         *     <TextInputMask
         *       {...props}
         *       mask="+[00] [000] [000] [000]"
         *     />
         *   }
         * />
         * ```
         */
        render?: (props: RenderProps) => ReactNode;
        /**
         * Value of the text input.
         */
        value?: string;
        /**
         * Pass `fontSize` prop to modify the font size inside `TextInput`.
         * Pass `height` prop to set `TextInput` height.
         * Pass `backgroundColor` prop to set `TextInput` backgroundColor.
         */
        style?: StyleProp<TextStyle>;
        /**
         * Style of the entire Container
         */
        containerStyle?: StyleProp<ViewStyle>;
        /**
         * Style of the Input Container
         */
        inputContainerStyle?: StyleProp<ViewStyle>;
        /**
         * Style of the Input
         */
        inputStyle?: StyleProp<TextStyle>;
        /**
         * props for the stateLayer
         */
        stateLayerProps?: PropsWithoutRef<ViewProps>;
        /**
         * testID to be used on tests.
         */
        testID?: string;
    };

type TextInputHandles = Pick<
    NativeTextInput,
    'focus' | 'clear' | 'blur' | 'isFocused' | 'setNativeProps'
>;

const TextInput = forwardRef<TextInputHandles, Props>(
    (
        {
            variant = 'flat',
            size = 'md',
            disabled = false,
            error: errorProp = false,
            multiline = false,
            editable = true,
            required = false,
            maxFontSizeMultiplier = 15,
            supportingText,
            selectionColor: selectionColorProp,
            underlineColor: underlineColorProp,
            activeUnderlineColor: activeUnderlineColorProp,
            outlineColor: outlineColorProp,
            activeOutlineColor: activeOutlineColorProp,
            placeholderTextColor: placeholderTextColorProp,
            style,
            containerStyle,
            hovered = false,
            ...rest
        }: Props,
        ref,
    ) => {
        const isControlled = rest.value !== undefined;
        const validInputValue = isControlled ? rest.value : rest.defaultValue;

        const { View, HelperText } = useMolecules();

        const { current: labelAnimation } = useRef<Animated.Value>(
            new Animated.Value(validInputValue ? 0 : 1),
        );
        const { current: errorAnimation } = useRef<Animated.Value>(
            new Animated.Value(errorProp ? 1 : 0),
        );
        const [focused, setFocused] = useState<boolean>(false);
        const [placeholder, setPlaceholder] = useState<string | undefined>('');
        // Use value from props instead of local state when input is controlled
        const [value, onChangeValue] = useControlledValue({
            value: rest.value,
            defaultValue: rest.defaultValue,
            onChange: rest.onChangeText,
            disabled: !editable || disabled,
        });

        const styles = useComponentStyles(
            'TextInput',
            [
                style,
                {
                    selectionColor: selectionColorProp,
                    underlineColor: underlineColorProp,
                    activeUnderlineColor: activeUnderlineColorProp,
                    outlineColor: outlineColorProp,
                    activeOutlineColor: activeOutlineColorProp,
                    placeholderTextColor: placeholderTextColorProp,
                },
            ],
            {
                variant,
                states: {
                    errorDisabled: errorProp && disabled,
                    disabled,
                    errorFocusedAndHovered: errorProp && hovered && focused,
                    errorFocused: errorProp && focused,
                    errorHovered: errorProp && hovered,
                    hoveredAndFocused: hovered && focused,
                    hovered,
                    focused: focused,
                    error: !!errorProp,
                },
                size,
            },
        );

        const [labelLayout, setLabelLayout] = useState<{
            measured: boolean;
            width: number;
            height: number;
        }>({
            measured: false,
            width: 0,
            height: 0,
        });

        const timer = useRef<NodeJS.Timeout | undefined>();

        const root = useRef<NativeTextInput | undefined | null>();

        useImperativeHandle(
            ref,
            () => ({
                focus: () => root.current?.focus(),
                clear: () => root.current?.clear(),
                setNativeProps: (args: Object) => root.current?.setNativeProps(args),
                isFocused: () => root.current?.isFocused() || false,
                blur: () => root.current?.blur(),
                forceFocus: () => root.current?.focus(),
            }),
            [],
        );

        useEffect(() => {
            // When the input has an error, we wiggle the label and apply error styles
            if (errorProp) {
                // show error
                Animated.timing(errorAnimation, {
                    toValue: 1,
                    duration: FOCUS_ANIMATION_DURATION * (styles.animationScale || 1),
                    // To prevent this - https://github.com/callstack/react-native-paper/issues/941
                    useNativeDriver: true,
                }).start();
            } else {
                // hide error
                {
                    Animated.timing(errorAnimation, {
                        toValue: 0,
                        duration: BLUR_ANIMATION_DURATION * (styles.animationScale || 1),
                        // To prevent this - https://github.com/callstack/react-native-paper/issues/941
                        useNativeDriver: true,
                    }).start();
                }
            }
        }, [errorProp, errorAnimation, styles]);

        useEffect(() => {
            // Show placeholder text only if the input is focused, or there's no label
            // We don't show placeholder if there's a label because the label acts as placeholder
            // When focused, the label moves up, so we can show a placeholder
            if (focused || !rest.label) {
                // Set the placeholder in a delay to offset the label animation
                // If we show it immediately, they'll overlap and look ugly
                timer.current = setTimeout(
                    () => setPlaceholder(rest.placeholder),
                    50,
                ) as unknown as NodeJS.Timeout;
            } else {
                // hidePlaceholder
                setPlaceholder('');
            }

            return () => {
                if (timer.current) {
                    clearTimeout(timer.current);
                }
            };
        }, [focused, rest.label, rest.placeholder]);

        useEffect(() => {
            // The label should be minimized if the text input is focused, or has text
            // In minimized mode, the label moves up and becomes small
            // workaround for animated regression for react native > 0.61
            // https://github.com/callstack/react-native-paper/pull/1440
            if (value || focused) {
                // minimize label
                Animated.timing(labelAnimation, {
                    toValue: 0,
                    duration: BLUR_ANIMATION_DURATION * (styles.animationScale || 1),
                    // To prevent this - https://github.com/callstack/react-native-paper/issues/941
                    useNativeDriver: true,
                }).start();
            } else {
                // restore label
                {
                    Animated.timing(labelAnimation, {
                        toValue: 1,
                        duration: FOCUS_ANIMATION_DURATION * (styles.animationScale || 1),
                        // To prevent this - https://github.com/callstack/react-native-paper/issues/941
                        useNativeDriver: true,
                    }).start();
                }
            }
        }, [focused, value, labelAnimation, styles]);

        const handleFocus = useCallback(
            (args: any) => {
                if (disabled || !editable) {
                    return;
                }

                setFocused(true);

                rest.onFocus?.(args);
            },
            [disabled, editable, rest],
        );

        const handleBlur = useCallback(
            (args: Object) => {
                if (!editable) {
                    return;
                }

                setFocused(false);
                rest.onBlur?.(args);
            },
            [editable, rest],
        );

        const handleLayoutAnimatedText = useCallback((e: LayoutChangeEvent) => {
            setLabelLayout({
                width: e.nativeEvent.layout.width,
                height: e.nativeEvent.layout.height,
                measured: true,
            });
        }, []);

        const forceFocus = useCallback(() => root.current?.focus(), []);

        const parentState = useMemo(
            () => ({
                labelAnimation,
                errorAnimation,
                focused,
                placeholder,
                value,
                labelLayout,
            }),
            [errorAnimation, focused, labelLayout, labelAnimation, placeholder, value],
        );

        return (
            <View style={containerStyle}>
                <TextInputBase
                    componentStyles={styles}
                    variant={variant}
                    size={size}
                    disabled={disabled}
                    error={errorProp}
                    multiline={multiline}
                    editable={editable}
                    required={required}
                    {...rest}
                    value={value}
                    parentState={parentState}
                    innerRef={ref => {
                        root.current = ref;
                    }}
                    onFocus={handleFocus}
                    forceFocus={forceFocus}
                    onBlur={handleBlur}
                    onChangeText={onChangeValue}
                    onLayoutAnimatedText={handleLayoutAnimatedText}
                    maxFontSizeMultiplier={maxFontSizeMultiplier}
                />
                <>
                    {(supportingText || required) && (
                        <HelperText
                            variant={errorProp ? 'error' : 'info'}
                            style={styles.supportingText}>
                            {supportingText || '*required'}
                        </HelperText>
                    )}
                </>
            </View>
        );
    },
);

export default memo(withActionState(TextInput));
