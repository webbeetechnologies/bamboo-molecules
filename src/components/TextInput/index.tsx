export { default as TextInput, Props as TextInputProps } from './TextInput';
export { defaultStyles as textInputStyles } from './utils';
