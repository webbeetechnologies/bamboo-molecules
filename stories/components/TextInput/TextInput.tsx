import { TextInput, TextInputProps } from '../../../src/components';

export type Props = TextInputProps;

export const Example = (props: Props) => {
    return <TextInput {...(props as any)} />;
};
