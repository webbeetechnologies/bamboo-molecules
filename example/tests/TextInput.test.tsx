import { Example as TextInput } from '../stories/components/TextInput/TextInput';
import { renderWithWrapper } from '../testHelpers';
import { Platform } from 'react-native';

it('renders text button by default', () => {
    const tree = renderWithWrapper(<TextInput />).toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders button with flat variant', () => {
    const tree = renderWithWrapper(<TextInput variant="flat" />).toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders button with outlined variant', () => {
    const tree = renderWithWrapper(<TextInput variant="outlined" />).toJSON();

    expect(tree).toMatchSnapshot();
});

it('renders button with left element', () => {
    jest.doMock('../stories/components/Icon/Icon', () => {
        return jest.fn(args => JSON.stringify(args));
    });

    const Icon = require('../stories/components/Icon/Icon');

    const { getByTestId, toJSON } = renderWithWrapper(
        <TextInput left={<Icon name="magnify" type="material-community" />} />,
    );

    expect(getByTestId('text-input-left-element')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
});

it('renders button with right element', () => {
    jest.doMock('../stories/components/Icon/Icon', () => {
        return jest.fn(args => JSON.stringify(args));
    });

    const Icon = require('../stories/components/Icon/Icon');

    const { getByTestId, toJSON } = renderWithWrapper(
        <TextInput right={<Icon name="magnify" type="material-community" />} />,
    );

    expect(getByTestId('text-input-right-element')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
});

it('correctly applies height to multiline Outline TextInput', () => {
    const { toJSON } = renderWithWrapper(
        <TextInput
            variant="outlined"
            label="Outline Input"
            placeholder="Type Something"
            value={'Some test value'}
            multiline
        />,
    );

    expect(toJSON()).toMatchSnapshot();
});

it('contains patch spacing for flat input when ios and multiline', () => {
    Platform.OS = 'ios';
    const onChangeTextMock = jest.fn();

    const { getByTestId } = renderWithWrapper(
        <TextInput
            label="Flat input"
            multiline
            placeholder="Type something"
            value={'Some test value'}
            onChangeText={onChangeTextMock}
        />,
    );

    expect(getByTestId('patch-container')).toBeTruthy();
});
