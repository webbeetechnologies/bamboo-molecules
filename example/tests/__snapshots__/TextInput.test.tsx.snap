// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly applies height to multiline Outline TextInput 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(255, 251, 254, 1)",
        "borderRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "backgroundColor": "rgba(255, 251, 254, 1)",
          "borderColor": "rgba(121, 116, 126, 1)",
          "borderRadius": 4,
          "borderWidth": 1,
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="text-input-outline"
    />
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <View
        collapsable={false}
        onLayout={[Function]}
        pointerEvents="none"
        style={
          Object {
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 0.25,
              },
            ],
            "zIndex": 3,
          }
        }
      >
        <Text
          collapsable={false}
          maxFontSizeMultiplier={15}
          numberOfLines={1}
          onLayout={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(255, 251, 254, 1)",
              "color": "rgba(73, 69, 79, 1)",
              "fontSize": 16,
              "fontWeight": "400",
              "left": 0,
              "lineHeight": 24,
              "marginLeft": -4,
              "opacity": 1,
              "paddingHorizontal": 4,
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
                Object {
                  "scale": 0.75,
                },
              ],
            }
          }
          testID="text-input-label-active"
        >
          Outline Input
        </Text>
        <Text
          collapsable={false}
          maxFontSizeMultiplier={15}
          numberOfLines={1}
          style={
            Object {
              "color": "rgba(73, 69, 79, 1)",
              "fontSize": 16,
              "fontWeight": "400",
              "left": 0,
              "lineHeight": 24,
              "marginLeft": -4,
              "opacity": 0,
              "paddingHorizontal": 4,
              "position": "absolute",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
                Object {
                  "scale": 0.75,
                },
              ],
            }
          }
          testID="text-input-label-inactive"
        >
          Outline Input
        </Text>
      </View>
      <TextInput
        editable={true}
        maxFontSizeMultiplier={15}
        multiline={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder=""
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 4,
              "paddingTop": 4,
            },
            Object {},
            Object {
              "paddingTop": 12,
            },
            Object {
              "textAlign": "left",
              "textAlignVertical": "top",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-outlined"
        underlineColorAndroid="transparent"
        value="Some test value"
      />
    </View>
  </View>
</View>
`;

exports[`renders button with flat variant 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <TextInput
        editable={true}
        maxFontSizeMultiplier={15}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 8,
              "paddingTop": 24,
            },
            Object {
              "height": 56,
            },
            false,
            Object {
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;

exports[`renders button with left element 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        Object {
          "color": "rgba(73, 69, 79, 1)",
          "iconSize": 20,
          "justifyContent": "center",
          "marginLeft": -4,
          "marginRight": 12,
        }
      }
      testID="text-input-left-element"
    >
      {"name":"magnify","type":"material-community"}
    </View>
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <TextInput
        editable={true}
        maxFontSizeMultiplier={15}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 8,
              "paddingTop": 24,
            },
            Object {
              "height": 56,
            },
            false,
            Object {
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;

exports[`renders button with outlined variant 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(255, 251, 254, 1)",
        "borderRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "backgroundColor": "rgba(255, 251, 254, 1)",
          "borderColor": "rgba(121, 116, 126, 1)",
          "borderRadius": 4,
          "borderWidth": 1,
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="text-input-outline"
    />
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <TextInput
        editable={true}
        maxFontSizeMultiplier={15}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 4,
              "paddingTop": 4,
            },
            Object {
              "height": 56,
            },
            false,
            Object {
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-outlined"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;

exports[`renders button with right element 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <TextInput
        editable={true}
        maxFontSizeMultiplier={15}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 8,
              "paddingTop": 24,
            },
            Object {
              "height": 56,
            },
            false,
            Object {
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
      />
    </View>
    <View
      style={
        Object {
          "color": "rgba(73, 69, 79, 1)",
          "iconSize": 24,
          "justifyContent": "center",
          "marginLeft": 12,
          "marginRight": -4,
        }
      }
      testID="text-input-right-element"
    >
      {"name":"magnify","type":"material-community"}
    </View>
  </View>
</View>
`;

exports[`renders text button by default 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "animationScale": 1,
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "flexDirection": "row",
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      style={
        Object {
          "flexGrow": 1,
          "minHeight": 56,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <TextInput
        editable={true}
        maxFontSizeMultiplier={15}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="rgba(73, 69, 79, 1)"
        size="md"
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "flexGrow": 1,
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
              "paddingBottom": 8,
              "paddingTop": 24,
            },
            Object {
              "height": 56,
            },
            false,
            Object {
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;
